ci_watson/artifactory_helpers.py

```diff
@@ -124,6 +152,309 @@ def get_bigdata(*args):
+            actual_hdu = get_hdu(actual)
+            desired_hdu = get_hdu(desired)
+
+            # Working with FITS Binary table with header...
+            fdiff = HDUDiff(actual_hdu, desired_hdu, rtol=rtol, atol=atol,
+                             ignore_keywords=ignore_keywords,
+                             ignore_fields=ignore_fields)
+            creature_report += fdiff.report()
+            if not fdiff.identical:
+                # Only keep track of failed results which need to
+                # be used to replace the truth files (if OK).
+                updated_outputs.append((actual_name, desired_name))
+            if not fdiff.identical and all_okay:
+                all_okay = False
+        else:
+            # ASCII-based diff
+            with open(actual) as afile:
+                actual_lines = afile.readlines()
+            with open(desired) as dfile:
+                desired_lines = dfile.readlines()
+            udiff = unified_diff(actual_lines, desired_lines,
+                                 fromfile=actual, tofile=desired)
+
+            old_stdout = sys.stdout
+            udiffIO = StringIO()
+            sys.stdout = udiffIO
+            sys.stdout.writelines(udiff)
+            sys.stdout = old_stdout
+            udiff_report = udiffIO.getvalue()
+            creature_report += udiff_report
+            if len(udiff_report) > 2 and all_okay:
+                all_okay = False
+            if len(udiff_report) > 2:
+                # Only keep track of failed results which need to
+                # be used to replace the truth files (if OK).
+                updated_outputs.append((actual, desired))
+
+    if not all_okay and results_root is not None:
+        tree = os.path.join(results_root, input_loc,
+                        dt, testdir) + os.sep
+
+        # Write out JSON file to enable retention of different results
+        new_truths = [os.path.basename(i[1]) for i in updated_outputs]
+        for files, new_truth in zip(updated_outputs, new_truths):
+            print("Renaming {} as new 'truth' file: {}".format(
+                  files[0], new_truth))
+            shutil.move(files[0], new_truth)
+        log_pattern = [os.path.join(os.path.dirname(x), '*.log') for x in new_truths]
+        generate_upload_schema(pattern=new_truths + log_pattern,
+                       testname=testname,
+                       target= tree)
+
+    if not all_okay and raise_error:
+        raise AssertionError(os.linesep + creature_report)
+
+    return creature_report
+
+
+def get_hdu(filename):
+    """Return the HDU for the file and extension specified in the filename.
+
+       This routine expects the filename to be of the format:
+           <filename>.fits[extn]
+
+        For example, "jw99999-a3001_t1_nircam_f140m-maskbar_i2d.fits[hdrtab]"
+    """
+    froot, fextn = filename.split('[')
+    fextn = fextn.replace(']','')
+    fits_file = fits.open(froot)
```

Comment:
Will this file handler be closed and garbage collected automatically when it goes out of scope? Will it ever go out of scope? If not, we might run into the risk of "too many open files" error. Same question for `build_hdulist` below.